### PR TITLE
Redesign of the rangebased for loops body

### PR DESCRIPTION
**Pull request by ovdiiuv**
Adding support for break/continue statements and nested loops.

Fixes:#1019
Fixes:#1033

### CODE REVIEW

File: include/clad/Differentiator/ReverseModeVisitor.h [hunk 1]
```diff
@@ -412,6 +412,7 @@ namespace clad {
     StmtDiff VisitCaseStmt(const clang::CaseStmt* CS);
     StmtDiff VisitDefaultStmt(const clang::DefaultStmt* DS);
     DeclDiff<clang::VarDecl> DifferentiateVarDecl(const clang::VarDecl* VD);
+    DeclDiff<clang::VarDecl> DifferentiateVarDecl(const clang::VarDecl* VD, bool AddToBlock);
```
Comment (vgvassilev) [inline include/clad/Differentiator/ReverseModeVisitor.h:415]:
```suggestion
    DeclDiff<clang::VarDecl> DifferentiateVarDecl(const clang::VarDecl* VD, bool AddToBlock = true);
```

That may simplify your life and you won't need to define another overload below.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 2]
```diff
@@ -976,75 +976,85 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+    PopBreakContStmtHandler();
 
     StmtDiff storeLoop = StoreAndRestore(BuildDeclRef(LoopVDDiff.getDecl()));
     StmtDiff storeAdjLoop =
         StoreAndRestore(BuildDeclRef(LoopVDDiff.getDecl_dx()));
-
     addToCurrentBlock(BuildDeclStmt(LoopVDDiff.getDecl_dx()));
-    Expr* CounterIncrement = loopCounter.getCounterIncrement();
 
+    // Expr* CounterIncrement = loopCounter.getCounterIncrement();
```
Comment (PetroZarytskyi) [inline lib/Differentiator/ReverseModeVisitor.cpp:1057]: Please either uncomment or remove this line. The same goes for the line ``addToCurrentBlock(CounterIncrement);`` below.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 3]
```diff
@@ -2926,6 +2931,10 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
     if (derivedVDE)
       m_Variables.emplace(VDClone, derivedVDE);
 
+    if ((VD->getDeclName() != VDClone->getDeclName() ||
+         VD->getType() != VDClone->getType()))
+      m_DeclReplacements[VD] = VDClone;
+
```
Comment (PetroZarytskyi) [inline lib/Differentiator/ReverseModeVisitor.cpp:2937]: If replacements are now handled in ``DifferentiateVarDecl``, can we not do this in ``VisitDeclStmt``?

**Comment by PetroZarytskyi** [comment]
The PR looks good to me. Great work!

**Pull request merged**